p2p/net/swarm/swarm_dial.go

```diff
@@ -35,13 +35,6 @@ const maxAddressResolution = 32
 //                                                             retry dialAttempt x
 
 var (
-	// ErrDialBackoff is returned by the backoff code when a given peer has
-	// been dialed too frequently
-	ErrDialBackoff = errors.New("dial backoff")
```

Comment:
`ErrNoGoodAddresses`

Reply:
It looks like people are explicitly testing for this error, and by removing it we'll break their builds: https://github.com/search?q=ErrDialBackoff&type=code.

When we make breaking changes, they need to be well motivated. Is that the case here?

Reply:
>When we make breaking changes, they need to be well motivated. Is that the case here?

Kinda. Admittedly, my aim here was only to simplify implementation and not improve user experience. But *all* those [checks]([github.com/search?q=ErrDialBackoff&type=code](https://github.com/search?q=ErrDialBackoff&type=code))  are broken because the errors returned from Connect do not satisfy either `err == ErrDialBackoff` or `errors.Is(err, ErrDialBackoff)`. 

The DialError returned from swarm.Connect is not very nice to use. This is something that I'd mentally put off for tackling with https://github.com/libp2p/specs/issues/479. Happy to take that up before this. 

From within swarm this test succeeds!: 
```go
func TestDialError(t *testing.T) {
	de := &DialError{
		Peer: "",
		DialErrors: []TransportError{
			{
				Address: ma.StringCast("/ip4/1.2.3.4/tcp/1234"),
				Cause:   ErrDialBackoff,
			},
		},
		Cause:   nil,
		Skipped: 0,
	}
	var err error = de
	if err == ErrDialBackoff {
		t.Fatalf("didn't expect backoff to be equal")
	}
	if errors.Is(err, ErrDialBackoff) {
		t.Fatalf("dial error is not backoff")
	}
}
```

Reply:
We should return `ErrDialBackoff` if we don't have any addresses we can use because some address we could have used is in a backoff state. This is much better error for a user than a "no good address" error. No good address would at least be confusing to users.

> From within swarm this test succeeds!:

It sound like we should add a test like this that fails and then fix it :)

Reply:
I have fixed this to keep the current behaviour.